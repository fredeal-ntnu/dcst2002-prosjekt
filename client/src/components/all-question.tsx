import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Column, Row, Form } from '../widgets';
import service, { Question, Tag } from '../service';

export class AllQuestions extends Component {
    questions: Question[] = [];
    filter: string = 'All'; // State to manage the filter type
    search = '';
    tags: Tag[] = [];
  
    render() {
      return (
        <>   
          <Card title='All Questions'>
            <Row>
                <Column width={2}>
                <Form.Select value={this.filter} onChange={this.handleFilterChange}>
                    <option value="all">All Questions</option>
                    <option value="popular">Most Popular</option>
                    <option value="unanswered">Unanswered</option>
                </Form.Select>
                </Column>
                <Column width={10}>
                    <Form.Input
                    type='text' 
                    value={this.search} 
                    onChange={(event) => this.search = event.currentTarget.value}
                    />
                </Column>
            </Row>

               
            {/* <Row>
            {this.tags.map((tag) => (
                        <Column>
                            <Form.Label>
                                {tag.name}
                                <input type='checkbox'
                                    value={tag.tag_id}
                                    onChange={() => {'vet ikke'}}
                                />
                            
                            </Form.Label>
                        
                        </Column>
                    ))}
            </Row> */}
          </Card>
          {this.questions
                .filter((question) => (question.title.toLowerCase().includes(this.search.toLowerCase())))
                .map((question, i) => (
                    <Card key={i} title={question.title}>
                        <Row>
                            <Column width={2}>{question.text}</Column>
                        </Row>
                    </Card>
                ))
                }
        </>
      );
    }

    handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        this.filter = event.target.value;
        this.loadQuestions(); // Call a method to load questions based on the selected filter
      };
  
    // Method to load questions based on the current filter
    loadQuestions() {
      // Add logic here to fetch and filter questions accordingly
      // This is a placeholder for whatever your service methods might be
      switch (this.filter) {
        case 'All':
          // Fetch all questions
          break;
        case 'Popular':
          // Fetch most popular questions
          break;
        case 'Unanswered':
          // Fetch unanswered questions
          break;
        default:
          // Handle default case
      }
    }
  
    // Lifecycle method to load questions when the component mounts
    mounted() {
      this.loadQuestions();

      service
        .getAllTags()
        .then((tags) => (this.tags = tags))
        .catch((error) => Alert.danger(error.message));  
      
      service
        .getAllQuestions()
        .then((questions) => (this.questions = questions))
        .catch((error) => Alert.danger(error.message));

  }
}