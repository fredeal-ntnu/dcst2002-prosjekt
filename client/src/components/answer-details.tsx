import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, Button, Form } from '../widgets';
import service, { Answer, AnswerComment, User } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();

export class AnswerDetails extends Component<{ match: { params: { id: number } } }> {
  answerComments: AnswerComment[] = [];
  answer: Answer = {
    answer_id: 0,
    text: '',
    confirmed_answer: 0,
    last_updated: new Date(),
    question_id: 0,
    user_id: 0,
  };
  answerComment: AnswerComment = { answer_comment_id: 0, text: '', answer_id: 0, user_id: 0 };
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  connectedUser: number = 0;

  render() {
    return (
      <>
        <Card title="Answer">
          <Row>
            <Column width={10}>{this.answer.text}</Column>
          </Row>
        </Card>
        <Card title="Comments">
          {this.answerComments.map((answerComment) => (
            <Row key={answerComment.answer_comment_id}>
              <Column width={10}>
                {answerComment.text}
                <Column>
                  <Button.Success
                    onClick={() =>
                      history.push(
                        '/questions/' +
                          this.answer.question_id +
                          '/answers/' +
                          answerComment.answer_id +
                          '/comments/' +
                          answerComment.answer_comment_id +
                          '/edit',
                      )
                    }
                  >
                    Edit
                  </Button.Success>
                </Column>
              </Column>
            </Row>
          ))}
        </Card>

        <Card title="Add Comment">
          <Form.Textarea
            placeholder="Add comment"
            type="text"
            value={this.answerComment.text}
            onChange={(event) => (this.answerComment.text = event.currentTarget.value)}
            maxLength={1000}
            style={{ width: '500px' }}
          />
          <Column>
            <Button.Success onClick={this.addAnswerComment}>Add Comment</Button.Success>
          </Column>
        </Card>
      </>
    );
  }

  mounted() {
    service

      //Get logged in user
      .getMe()
      .then((user) => {
        this.user = user;
        this.connectedUser = this.user.user_id;
      })
      .catch((error) => {
        console.error(error.message);
        history.push('/');
        alert('You must be logged in to access this page');
      });

    service
      .getAnswerById(this.props.match.params.id)
      .then((answer) => {
        this.answer = answer;
        return service.getAnswerCommentsForAnswer(this.answer.answer_id);
      })
      .then((answerComments) => (this.answerComments = answerComments))
      .catch((error: Error) => console.error('Error getting answer: ' + error.message));
  }

  addAnswerComment() {
    
    if(this.connectedUser) {
      service
      .createAnswerComment(this.answerComment.text, this.answer.answer_id, this.connectedUser)
      .then(() => this.mounted())
      .then(() => (this.answerComment.text = ''))
      .catch((error) => console.error('Error adding answer comment: ' + error.message));
    }else return
    
  }
}
