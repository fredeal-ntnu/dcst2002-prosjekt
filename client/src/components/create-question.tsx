import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, SideMenu, Button, Alert, Form } from '../widgets';
import service, { Tag } from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class CreateQuestion extends Component {
  tags: Tag[] = [];
  selectedTags: number[] = [];
  title = '';
  text = '';
  user_id = 0;
  connectedUser: number = 0;

  render() {
    return (
      <>
        <Row>
          <Column width={3}>
            <SideMenu
              header="Menu"
              items={[
                { label: 'Questions', to: '/questions' },
                { label: 'My Questions', to: 'myquestions' },
                { label: 'My Favourite Answers', to: 'favourites' },
                { label: 'New Question', to: 'createquestion' },
                { label: 'Tags', to: '/tags' },
              ]}
            />
          </Column>
          <Card title="Ask a Question">
            <Row>
              <Column width={2}>
                <Form.Label>Title:</Form.Label>
              </Column>
              <Column>
                <Form.Input
                  id="createTitle"
                  type="text"
                  value={this.title}
                  onChange={(event) => (this.title = event.currentTarget.value)}
                  maxLength={255}
                />
              </Column>
            </Row>
            <br />
            <Row>
              <Column width={2}>
                <Form.Label>Text:</Form.Label>
              </Column>
              <Column>
                <Form.Textarea
                  id="createText"
                  placeholder="Text"
                  value={this.text}
                  onChange={(event) => (this.text = event.currentTarget.value)}
                  rows={5}
                  maxLength={1000}
                />
              </Column>
            </Row>
            <br />
            <Row>
              <Column width={2}>
                <Form.Label>Tags:</Form.Label>
              </Column>
              <Column>
                <Row>
                  {this.tags.map((tag) => (
                    <Column key={tag.tag_id}>
                      <Form.Label>
                        {tag.name}
                        <Form.Checkbox
                          type="checkbox"
                          value={tag.tag_id}
                              checked={this.selectedTags.includes(tag.tag_id)}
                          onChange={(event) => {
                            this.handleCheckboxChange(event);
                          }}
                        />
                      </Form.Label>
                    </Column>
                  ))}
                </Row>
              </Column>
            </Row>
            <br />
            <Row>
              <Column>
                <Button.Success
                  onClick={() => {
                    this.handleAddQuestion();
                  }}
                >
                  AskMorgan
                </Button.Success>
              </Column>
            </Row>
          </Card>
        </Row>
      </>
    );
  }

  mounted() {
    service
      .getMe()
      .then((user) => {
        this.user_id = user.user_id;
        this.connectedUser = this.user_id;
      })
      .catch((error) => {
        console.error(error.message);
        history.push('/');
        alert('You must be logged in to create a question');
      });

    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .catch((error) => console.error(error.message));
  }

  handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let tagId = Number(event.target.value);

    if (event.target.checked) {
      this.selectedTags.push(tagId);
    } else {
      this.selectedTags = this.selectedTags.filter((id) => id !== tagId);
    }
  };

  handleAddQuestion = async () => {
    if (this.selectedTags.length == 0) {
      //error
      alert('You must select at least one tag');
      return null;
    } else {
      var question_id = await service.createQuestion(this.title, this.text, this.user_id);
    }

    // For each selected tag, create a new relation in the Tag_question_relation tabl
    this.selectedTags.forEach((tag_id) => {
      service.createTagQuestionRelation(tag_id, question_id);
    });

    // Redirect to the question page
    history.push(`/questions/${question_id}`);
  };
}
