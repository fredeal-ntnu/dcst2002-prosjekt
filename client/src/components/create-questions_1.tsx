import * as React from 'react';
import { Component } from 'react-simplified';
import { Alert, Form, Card, Row, Column, Button} from '.././widgets';
import { NavLink } from 'react-router-dom';
import service, { Question, Tag_Question_Relation, Tag } from '.././service';
import { createHashHistory } from 'history';

const history = createHashHistory(); // Use history.push(...) to programmatically change path, for instance after successfully saving a student
export class CreateQuestion extends Component {
    tags: Tag[] = [];
    title = "";
    text = "";
    selectedTags: number[] = []; 
  

  render() {
    return (
          <Row>
            <Form.Label>Ask a question</Form.Label>
                <Form.Input
                    type='text'
                    placeholder='Tittel'
                    value={this.title}
                    onChange={(event) => (this.title = event.currentTarget.value)}
                />
                <Form.Input
                    type='text'
                    placeholder='Text'
                    value={this.text}
                    onChange={(event) => (this.text = event.currentTarget.value)}
                />
                <Row>
                    {this.tags.map((tag) => (
                        <Column>
                            <Form.Label>
                                {tag.name}
                                <input type='checkbox'
                                    value={tag.tag_id}
                                    onChange={(event) => {this.handleCheckboxChange(event)}}
                                />
                            
                            </Form.Label>
                        
                        </Column>
                    ))}
                   
                </Row>
                <Button.Success onClick={() => {this.handleAddQuestion()}}>Create</Button.Success>                    
          </Row>
    );
  }

  mounted() {
        
    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .catch((error) => Alert.danger(error.message));
      console.log(this.tags)
  
  }

  handleCheckboxChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    let tagId = Number(event.target.value);

    if (event.target.checked) {
        this.selectedTags.push(tagId);
    } else {
        this.selectedTags = this.selectedTags.filter(id => id !== tagId);
    }
}


handleAddQuestion = async () => {
    console.log(typeof(this.selectedTags))
    const question_id = await service.createQuestion(this.title, this.text);

    // For each selected tag, create a new relation in the Tag_question_relation tabl
    this.selectedTags.forEach(tag_id => {
        service.createTagQuestionRelation(tag_id, question_id);
    });
    
    // Redirect to the question page
    history.push(`/questions/${question_id}`);
}


}







/*

createquestion{
  hent ut alle tagger og display på siden i form av checkbox

  insert tittel og text fra bruker
  send til tabell
  få ut result.question id

  bruk question.id til å loope igjennom alle tagger som vi har hentet, ved bruk av tag.id som vi kan sjekke med if checked.
  for alle tags funnet med den question.id, insert question.id og tag.id i tag_question_relation


}

*/