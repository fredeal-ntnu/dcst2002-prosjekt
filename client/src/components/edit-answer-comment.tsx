import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Alert, Row, Column, Button, Form} from '../widgets';
import service, {AnswerComment, User} from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();
// let history = useHistory();
export class EditAnswerComment extends Component<{ match: { params: { id: number } } }> {
  answerComments: AnswerComment[] = [];
  answerComment: AnswerComment = {answer_comment_id: 0, text: '', answer_id: 0, user_id: 0};
  user: User = { user_id: 0, google_id: '', username: '', email: '' };
  connectedUser: number = 0;
  


  render() {
    return(
      <>
      <Card title="Edit Comment">
        <Row>
          <Column width={10}>
          <Form.Textarea
                type="text"
                value={this.answerComment.text}
                onChange={(event) => (this.answerComment.text = event.currentTarget.value)}
                rows={5}
              />
          </Column>
        </Row>
        <Row>
          <Column><Button.Success onClick={this.save}>Save</Button.Success></Column>
          <Column><Button.Danger onClick={this.delete}>Delete</Button.Danger></Column>
        </Row>
      </Card>
      </>
    )
    
  }

  mounted() {
    service.getMe().then((user) => {
      this.user = user;
      this.connectedUser = user.user_id;
    });


    service.getAnswerCommentById(this.props.match.params.id)
    .then((answerComment) => (this.answerComment = answerComment))
    .catch((error: Error) => Alert.danger('Error getting answer comment: ' + error.message));
  }

  save() {
    if(this.connectedUser == this.answerComment.user_id){
    service
      .updateAnswerComment(this.answerComment)
      .then(() => history.goBack())
      .catch((error) => Alert.danger('Error saving answer comment: ' + error.message));
  }else{
    Alert.danger('You are not the owner of this comment');
  }
}

  delete() {
    if(this.connectedUser == this.answerComment.user_id){
    service
    .deleteAnswerComment(this.answerComment.answer_comment_id)
    .then(() => history.goBack())
    .catch((error) => Alert.danger('Error deleting answer comment: ' + error.message));
  }else{
    Alert.danger('You are not the owner of this comment');
  }
}

}
