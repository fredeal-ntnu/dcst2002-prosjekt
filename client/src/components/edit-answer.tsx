import * as React from 'react';
import { Component } from 'react-simplified';
import { Card, Row, Column, Button, Alert, Form } from '../widgets';
import service, { Answer, User} from '../service';
import { createHashHistory } from 'history';

const history = createHashHistory();

export class EditAnswer extends Component<{ match: { params: { id: number } } }> {
  answer: Answer = { answer_id: 0, text: '', confirmed_answer: 0, last_updated: new Date(), question_id: 0, user_id: 0};
  user: User = { user_id: 0, google_id: '', username: '', email: ''};
  connectedUser: number = 0;

  render() {
    return (
      <>
      <Row>
        <Card title="Edit Answer">
          <Row>
            <Column>
              <Form.Textarea
                placeholder="Edit answer"
                type="text"
                value={this.answer.text}
                onChange={(event) => (this.answer.text = event.currentTarget.value)}
              />
            </Column>
          </Row>
        </Card>
      </Row>
      <Row>
        <Column width={6}>
          <Button.Success onClick={this.save}>Save</Button.Success>
        </Column>
        <Column width={6}>
          <Button.Danger onClick={this.delete}>Delete</Button.Danger>
        </Column>
      </Row>
    </>
    )
  }

  mounted() {
    service.getMe()
    .then((user) => {
      this.user.user_id = user.user_id
      this.connectedUser = this.user.user_id;
    })
    .catch((error)=>{
      console.error(error.message)
      history.push('/')
      alert('You must be logged in to edit answer')
    })

    service
      .getAnswerById(this.props.match.params.id)
      .then((answer) => (this.answer = answer))
      .catch((error: Error) => console.error('Error getting answer: ' + error.message));
  }

  save() {
    if(this.answer.user_id == this.connectedUser) {
    service
      .updateAnswer(this.answer)
      .then(() => history.push('/questions/' + this.answer.question_id))
      .catch((error) => console.error('Error saving answer: ' + error.message));
  }else{
    history.push('/questions/' + this.answer.question_id)
    console.error('You are not the author of this answer')
  }
}

  delete() {
    if(this.answer.user_id == this.connectedUser)  {
      service
      .deleteAnswer(this.props.match.params.id)
      .then(() => history.push('/questions/' + this.answer.question_id))
      .catch((error) => console.error('Error deleting answer: ' + error.message));
    }else{
      history.push('/questions/' + this.answer.question_id)
      console.error('You are not the author of this answer')
    }
   
  }
}


