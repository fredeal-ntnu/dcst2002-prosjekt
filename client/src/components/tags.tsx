import * as React from 'react';
import { Component } from 'react-simplified';
import { Form, Card, Alert, MainCard, SideMenu, Row, Column, Button, RadioRow, QuestionCard } from '../widgets';
import service, { Question, Tag, Tag_Question_Relation } from '../service';
import { NavLink } from 'react-router-dom';
import { EyeIcon } from '../icons';

type TagWithCount = {
  tag_id: number;
  name: string;
  questions: number;
};

export class Tags extends Component {
    questions: Question[] = [];
    tags: TagWithCount[] = [];
    relations: Tag_Question_Relation[] = [];   
    questionSearch = '';
    tagSearch = '';
    sortOrder = 'none';
  
    render() {
        return (
            <Row>
            <Column width={3}>
                <SideMenu header='Menu'
                items={[
                  { label: "Questions", to: "/questions" },
                  { label: "Tags", to: "/tags" },
                  {label: "New Question", to: "createquestion"},
                  {label: "My Questions", to: "myquestion"}
                ]}/>
              <SideMenu header="Select Tags">
              <div style={{ padding: '10px', paddingBottom: '0px'}}>
                <Form.Select value={this.sortOrder} onChange={this.handleFilterChange}>
                  <option value="all">Sort tags...</option>
                  <option value="popularity">Popularity</option>
                  <option value="a-z">A-Z</option>
                  <option value="z-a">Z-A</option>
                </Form.Select>
                </div>
                <div style={{ padding: '10px'}}>
                  <Form.Input
                    type='text' 
                    placeholder={`Search`}
                    value={this.tagSearch} 
                    onChange={(event) => this.tagSearch = event.currentTarget.value}
                  />
                </div>              
                <ul className="list-group list-group-flush">
                  {this.tags
                    .filter((tag) => (tag.name.toLowerCase().includes(this.tagSearch.toLowerCase())))
                    .map((tag, index) => (
                      <li key={index} className="list-group-item">
                        <Button.Light key={tag.tag_id} onClick={() => this.showQuestionsByTag(tag.tag_id)}>
                          {tag.name}{' '}{tag.questions}
                        </Button.Light>
                      </li>
                    ))
                  }   
                </ul>
              </SideMenu>
            </Column>

            {/* Main content */}
            <Column>
              <Card title="Top Questions" width='600px'><br />
                <Row>
                  <Column>
                    <Form.Input
                    type='text' 
                    placeholder={`Search`}
                    value={this.questionSearch} 
                    onChange={(event) => this.questionSearch = event.currentTarget.value}/>
                  </Column>
                </Row>
              </Card>
                {this.questions
                  .filter((question) => (question.title.toLowerCase().includes(this.questionSearch.toLowerCase())))
                  .map((question, i) => (
                    <QuestionCard key={i} question={question} />
                  ))
                }
            </Column>
            </Row>
        );
      }

  showQuestionsByTag(tag_id: number) {
    service
      .getQuestionsByTagId(tag_id)
      .then((relations) => (this.relations = relations))
      .then(() => {
        this.questions = [];
        this.relations.map((relation) => {
          service
            .getQuestion(relation.question_id)
            .then((question) => this.questions.push(question))
            .catch((error) => Alert.danger(error.message));
        });
      })
      .catch((error) => Alert.danger(error.message));
  }

    handleFilterChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
        this.sortOrder = event.target.value;
        console.log(this.sortOrder);
        this.handleSortChange() // Call a method to load questions based on the selected filter
      };

    handleSortChange() {        
        switch (this.sortOrder) {
            // må ha alle her, hvordan gjør man det
            case 'popularity':
                this.tags.sort((a, b) => b.questions - a.questions);
                break;
            case 'a-z':
                this.tags.sort((a, b) => a.name.localeCompare(b.name));
                break;
            case 'z-a':
                this.tags.sort((a, b) => b.name.localeCompare(a.name));
                break;
            default:
                this.tags.sort((a, b) => a.tag_id - b.tag_id);
                break;
        }
      }
  
    mounted() {
      service
        .getAllQuestions()
        .then((questions) => (this.questions = questions));

    service
      .getAllTags()
      .then((tags) => (this.tags = tags))
      .then(() => {
        this.tags.map((tag, i) => {
          service
            .getQuestionsByTagId(tag.tag_id)
            .then((relations) => (this.relations = relations))
            .then(() => {
              this.tags[i].questions = this.relations.length;
            })
            .catch((error) => Alert.danger(error.message));
        });
      })
      .catch((error) => Alert.danger(error.message));
  }
}
