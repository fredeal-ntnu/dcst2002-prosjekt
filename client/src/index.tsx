import * as React from 'react';
import { createRoot } from 'react-dom/client';
import { Component } from 'react-simplified';
import { HashRouter, Route } from 'react-router-dom';
import { NavBar, Card, Alert, CardQuestions, CardHome, Row, Column } from './widgets';
import service, { Question } from './service';
// import { TaskList, TaskDetails, TaskEdit, TaskNew } from './task-components';


class Menu extends Component {
  render() {
    return (
      <NavBar brand="askMorgan">
        <NavBar.Search/>
        <NavBar.Link to="/login">Log in</NavBar.Link>
        <NavBar.Link to="/signup">Sign up</NavBar.Link>
      </NavBar>
    );
  }
}

class Home extends Component {
  questions: Question[] = [];

  render() {
    return (
      <Card title="">
        <div className="row">
          <CardHome header="Public"
            items={[{ label: "Questions", to: "/questions" }, { label: "Tags", to: "/tags" }]}/>
          <CardQuestions header="Top Questions">
          {this.questions.map((question) => (
            <Row key={question.question_id}>
              {question.title}<br></br>{question.text}
            </Row>
          ))}
        </CardQuestions>
          
        </div> 
      </Card>
    );
  }


  createQuestion() {
    service.createQuestion().then((questions) => (this.questions = questions));
  }



    mounted() {
      service.getTopFiveQuestions().then((questions) => (this.questions = questions));
    }
}


let root = document.getElementById('root');
if (root)
  createRoot(root).render(
    <HashRouter>
      <div>
        <Alert />
        <Menu />
        <Route exact path="/" component={Home} />
        {/* <Route exact path="/signin" component={login} /> */}
        {/* <Route exact path="/signup" component={signup} /> */}
        {/* <Route exact path="/signup" component={questions} /> */}
        {/* <Route exact path="/signup" component={tags} /> */}
      </div>
    </HashRouter>,
  );
