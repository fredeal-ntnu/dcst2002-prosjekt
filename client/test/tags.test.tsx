import React from 'react';
import { shallow } from 'enzyme';
import { Tags } from 'src/components/tags'; // Adjust the import path as needed
import { Column } from '../src/widgets';

jest.mock('../src/service', () => {
  class Service {
    getAllQuestions() {
      return new Promise((resolve, reject) => {
        resolve([
          {
            question_id: 1,
            title: 'test',
            description: 'test',
            user_id: 1,
            username: 'test',
            date: 'test',
            score: 1,
          },
        ]);
      });
    }
    getAllQuestionsByTagId(tag_id: number) {
      return new Promise((resolve, reject) => {
        resolve([
          {
            question_id: 1,
            title: 'test',
            description: 'test',
            user_id: 1,
            username: 'test',
            date: 'test',
            score: 1,
          },
        ]);
      });
    }
    getAllTags() {
      return new Promise((resolve, reject) => {
        resolve([{ tag_id: 1, name: 'test' }]);
      });
    }
  }
  return new Service();
}); // Mock the service module

describe('Site renders', () => {
  test('Site renders correct', (done) => {
    const wrapper = shallow(<Tags />);
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();
      done();
    });
  });
});

describe('site functionality', () => {
  test('search onchange on first search works', (done) => {
    const wrapper = shallow(<Tags />);
    console.log(wrapper.debug());

    // Find the specific input node using the appropriate selector
    const inputNode = wrapper.find('FormInput');

    // Simulate the change on the first node in the array
    inputNode.first().simulate('change', { currentTarget: { value: 'test' } });
    // Update the wrapper to reflect the simulated change
    wrapper.update();

    // Assert the changes with the snapshot
    expect(wrapper).toMatchSnapshot();

    // Call done to finish the test
    done();
});

test('search onchange works for second node', (done) => {
  const wrapper = shallow(<Tags />);
  console.log(wrapper.debug());

  // Find all nodes that match the selector
  const inputNodes = wrapper.find('FormInput');

  // Ensure that there are at least two nodes
  expect(inputNodes).toHaveLength(2);

  // Simulate the change on the second node in the array (index 1)
  inputNodes.at(1).simulate('change', { currentTarget: { value: 'test' } });

  // Update the wrapper to reflect the simulated change
  wrapper.update();

  // Assert the changes with the snapshot
  expect(wrapper).toMatchSnapshot();

  // Call done to finish the test
  done();
});
});

describe('switchcase', () => {
  const wrapper = shallow(<Tags />);

  test('switchcase popular', () => {
    wrapper.find('FormSelect').simulate('change', { target: { value: 'popularity' } });

    expect(wrapper).toMatchSnapshot();
  });

  test('switchcase a-z', () => {
    wrapper.find('FormSelect').simulate('change', { target: { value: 'a-z' } });
    expect(wrapper).toMatchSnapshot();
  });
  test('switchcase z-a', () => {
    wrapper.find('FormSelect').simulate('change', { target: { value: 'z-a' } });
    expect(wrapper).toMatchSnapshot();
  });
});
