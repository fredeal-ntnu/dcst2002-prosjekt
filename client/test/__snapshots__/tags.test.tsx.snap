// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Site renders Site renders correct 1`] = `
<Row>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "favourites",
          },
          {
            "label": "New Question",
            "to": "createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
    <SideMenu
      header="Select Tags"
    >
      <div
        style={
          {
            "padding": "10px",
            "paddingBottom": "0px",
          }
        }
      >
        <FormSelect
          onChange={[Function]}
          value="none"
        >
          <option
            value="all"
          >
            Sort tags...
          </option>
          <option
            value="popularity"
          >
            Popularity
          </option>
          <option
            value="a-z"
          >
            A-Z
          </option>
          <option
            value="z-a"
          >
            Z-A
          </option>
        </FormSelect>
      </div>
      <div
        style={
          {
            "padding": "10px",
          }
        }
      >
        <FormInput
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <ul
        className="list-group list-group-flush"
      >
        <li
          className="list-group-item"
          key="0"
        >
          <ButtonLight
            key="1"
            onClick={[Function]}
          >
            test
             
            1
          </ButtonLight>
        </li>
      </ul>
    </SideMenu>
  </Column>
  <Column>
    <Card
      title="Top Questions"
    >
      <br />
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value=""
          />
        </Column>
      </Row>
    </Card>
    <QuestionCard
      key="0"
      question={
        {
          "date": "test",
          "description": "test",
          "question_id": 1,
          "score": 1,
          "title": "test",
          "user_id": 1,
          "username": "test",
        }
      }
    />
  </Column>
</Row>
`;

exports[`switchcase switchcase a-z 1`] = `
<Row>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "favourites",
          },
          {
            "label": "New Question",
            "to": "createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
    <SideMenu
      header="Select Tags"
    >
      <div
        style={
          {
            "padding": "10px",
            "paddingBottom": "0px",
          }
        }
      >
        <FormSelect
          onChange={[Function]}
          value="a-z"
        >
          <option
            value="all"
          >
            Sort tags...
          </option>
          <option
            value="popularity"
          >
            Popularity
          </option>
          <option
            value="a-z"
          >
            A-Z
          </option>
          <option
            value="z-a"
          >
            Z-A
          </option>
        </FormSelect>
      </div>
      <div
        style={
          {
            "padding": "10px",
          }
        }
      >
        <FormInput
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <ul
        className="list-group list-group-flush"
      >
        <li
          className="list-group-item"
          key="0"
        >
          <ButtonLight
            key="1"
            onClick={[Function]}
          >
            test
             
            1
          </ButtonLight>
        </li>
      </ul>
    </SideMenu>
  </Column>
  <Column>
    <Card
      title="Top Questions"
    >
      <br />
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value=""
          />
        </Column>
      </Row>
    </Card>
    <QuestionCard
      key="0"
      question={
        {
          "date": "test",
          "description": "test",
          "question_id": 1,
          "score": 1,
          "title": "test",
          "user_id": 1,
          "username": "test",
        }
      }
    />
  </Column>
</Row>
`;

exports[`switchcase switchcase popular 1`] = `
<Row>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "favourites",
          },
          {
            "label": "New Question",
            "to": "createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
    <SideMenu
      header="Select Tags"
    >
      <div
        style={
          {
            "padding": "10px",
            "paddingBottom": "0px",
          }
        }
      >
        <FormSelect
          onChange={[Function]}
          value="popularity"
        >
          <option
            value="all"
          >
            Sort tags...
          </option>
          <option
            value="popularity"
          >
            Popularity
          </option>
          <option
            value="a-z"
          >
            A-Z
          </option>
          <option
            value="z-a"
          >
            Z-A
          </option>
        </FormSelect>
      </div>
      <div
        style={
          {
            "padding": "10px",
          }
        }
      >
        <FormInput
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <ul
        className="list-group list-group-flush"
      >
        <li
          className="list-group-item"
          key="0"
        >
          <ButtonLight
            key="1"
            onClick={[Function]}
          >
            test
             
            1
          </ButtonLight>
        </li>
      </ul>
    </SideMenu>
  </Column>
  <Column>
    <Card
      title="Top Questions"
    >
      <br />
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value=""
          />
        </Column>
      </Row>
    </Card>
    <QuestionCard
      key="0"
      question={
        {
          "date": "test",
          "description": "test",
          "question_id": 1,
          "score": 1,
          "title": "test",
          "user_id": 1,
          "username": "test",
        }
      }
    />
  </Column>
</Row>
`;

exports[`switchcase switchcase z-a 1`] = `
<Row>
  <Column
    width={3}
  >
    <SideMenu
      header="Menu"
      items={
        [
          {
            "label": "Questions",
            "to": "/questions",
          },
          {
            "label": "My Questions",
            "to": "myquestions",
          },
          {
            "label": "My Favourite Answers",
            "to": "favourites",
          },
          {
            "label": "New Question",
            "to": "createquestion",
          },
          {
            "label": "Tags",
            "to": "/tags",
          },
        ]
      }
    />
    <SideMenu
      header="Select Tags"
    >
      <div
        style={
          {
            "padding": "10px",
            "paddingBottom": "0px",
          }
        }
      >
        <FormSelect
          onChange={[Function]}
          value="z-a"
        >
          <option
            value="all"
          >
            Sort tags...
          </option>
          <option
            value="popularity"
          >
            Popularity
          </option>
          <option
            value="a-z"
          >
            A-Z
          </option>
          <option
            value="z-a"
          >
            Z-A
          </option>
        </FormSelect>
      </div>
      <div
        style={
          {
            "padding": "10px",
          }
        }
      >
        <FormInput
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
      </div>
      <ul
        className="list-group list-group-flush"
      >
        <li
          className="list-group-item"
          key="0"
        >
          <ButtonLight
            key="1"
            onClick={[Function]}
          >
            test
             
            1
          </ButtonLight>
        </li>
      </ul>
    </SideMenu>
  </Column>
  <Column>
    <Card
      title="Top Questions"
    >
      <br />
      <Row>
        <Column>
          <FormInput
            onChange={[Function]}
            placeholder="Search..."
            type="text"
            value=""
          />
        </Column>
      </Row>
    </Card>
    <QuestionCard
      key="0"
      question={
        {
          "date": "test",
          "description": "test",
          "question_id": 1,
          "score": 1,
          "title": "test",
          "user_id": 1,
          "username": "test",
        }
      }
    />
  </Column>
</Row>
`;
